### PR TITLE
All 3 algorithms work

### PR DESCRIPTION
**Pull request by cirillom**
**DOCUMENTATION NEEDED FOR EVERYTHING**

A* partitioned:
- not optimal;
- does not work if there is no direct connection between floors
- is the ugliest code of my life

 
Updates needed:
- Show number of steps to solution
- Show solution cost

Bugs:
- Video slider does not work while playing

### CODE REVIEW

File: src/maze_solver.py [hunk 1]
```diff
@@ -49,18 +50,29 @@ def solveAStar(
     print(path) # prints [(0, 0, 0), (0, 1, 0), (0, 2, 0)]
```
Comment (lucasgpulcinelli) [inline src/maze_solver.py]: This docstring does not reflect the actual usage now, i recommend removing the example and changing the documentation.

File: src/a_star.py [hunk 2]
```diff
@@ -39,13 +40,244 @@ def return_path(current_node):
...
+            paths = add_floor_to_steps(paths, start[0])
+            borders = add_floor_to_steps(borders, start[0])
+
+            path_step += paths
+            border_step += borders
+            path = paths[-1]
+            if path is not None:
+                end_calculated_destinations = [x[0] for x in paths_in_end_floor]
+                if destiny in end_calculated_destinations:
+                    end_path = paths_in_end_floor[end_calculated_destinations.index(destiny)][1]
```
Comment (lucasgpulcinelli) [inline src/a_star.py]: would it be possible to reorganize this code to use more "continues" in order to lower the amount of indentation and ease understanding?

Comment (lucasgpulcinelli) [reply src/a_star.py]: For instance, the major if-else in this block could be swapped with `if i >= len(start_euclidean_distances)` and a continue added at the end, which would make the code easier to reason about

File: src/a_star.py [hunk 3]
```diff
@@ -39,13 +40,244 @@ def return_path(current_node):
...
+                    border_step.append(None)
+                    return path_step, border_step
+
+                paths_in_start_floor.append((destiny, path))
+        else:
+            stop -= 1
+            if stop == 0:
+                break
+
+        if i < len(end_euclidean_distances):
```
Comment (lucasgpulcinelli) [inline src/a_star.py]: Also, is this code duplicated? It seems almost identical to the first if condition inside the loop, would it be possible to create a function for that or something in order to abstract what this code does?

File: src/a_star.py [hunk 4]
```diff
@@ -39,13 +40,244 @@ def return_path(current_node):
...
+    # what squares do we search
+    if allow_diagonal_movement:
+        adjacent_squares = ((0, -1), (0, 1), (-1, 0), (1, 0), (-1, -1), (-1, 1), (1, -1), (1, 1))
+        direction_cost = (horizontal_cost, horizontal_cost, horizontal_cost, horizontal_cost, diagonal_cost, diagonal_cost, diagonal_cost, diagonal_cost)
+        adjacent_square_pick_index = [0, 1, 2, 3, 4, 5, 6, 7]
+
+    # Loop until you find the end
+    while len(open_list) > 0:
+        #Randomize the order of the adjacent_squares_pick_index to avoid a decision making bias
+        random.shuffle(adjacent_square_pick_index)
```
Comment (lucasgpulcinelli) [inline src/a_star.py]: I don't understand why this random call is necessary, why would favouring any particular direction be a problem? The cost dictates the best path anyways

File: src/bfs.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+from collections import deque
+import numpy as np
+ 
+def isValid(vis, floor, row, col):
+    #Cell out of bounds
+    if (row < 0 or col < 0 or row >= 70 or col >= 102 or floor < 0 or floor >= 5):
```
Comment (lucasgpulcinelli) [inline src/bfs.py]: would it be possible not to hardcode values that work only for our specific images?

**Review by lucasgpulcinelli — CHANGES_REQUESTED** [review]
Aside from the fact that the video related changes should not have been in this branch (this WILL cause merge conflicts at some point), the code is mostly solid except for some messy code in the A* partitioned loop, an inconsistent docstring and a hardcoded value in BFS..

File: src/a_star.py (hunk 4 above)
Comment (cirillom) [reply src/a_star.py]: it's not a good idea, this is useless

File: src/a_star.py (hunk 2 above)
Comment (cirillom) [reply src/a_star.py]: the astar_partitioned code is utter trash (one of the worst piece of computer instructions i've ever done in my life)

i'm going to make it more readable

File: src/maze_solver.py (hunk 1 above)
Comment (cirillom) [reply src/maze_solver.py]: done

File: src/a_star.py (hunk 3 above)
Comment (cirillom) [reply src/a_star.py]: done

File: src/a_star.py (hunk 2 above)
Comment (cirillom) [reply src/a_star.py]: fixed it, hopefully is better

File: src/bfs.py (hunk 5 above)
Comment (cirillom) [reply src/bfs.py]: done

**Review by lucasgpulcinelli — APPROVED** [review]
LGTM

**Pull request merged**